feature-governance-impl/src/main/java/io/novafoundation/nova/feature_governance_impl/presentation/tindergov/cards/TinderGovCardsViewModel.kt

```diff
@@ -252,6 +273,28 @@ class TinderGovCardsViewModel(
             imageTintRes = if (items.isEmpty()) R.color.icon_inactive else R.color.chip_icon,
         )
     }
+
+    private fun observeReferendaAndAddToCards() {
+        combineToPair(interactor.observeReferendaAvailableToVote(this), basketFlow)
+            .onEachWithPrevious { old, new -> checkItemsRemovedFromBasket(old?.second.orEmpty(), new.second) }
+            .onEach { (referenda, basket) -> addNewReferendaToCards(referenda, basket) }
+            .launchIn(this)
+    }
+
+    private fun addNewReferendaToCards(referenda: List<ReferendumPreview>, basket: Map<ReferendumId, TinderGovBasketItem>) {
+        val currentReferendaIds = sortedReferendaFlow.value.map { it.id }.toSet()
+        val newReferenda = referenda.filter { it.id !in currentReferendaIds && it.id !in basket }
+
+        sortedReferendaFlow.value += newReferenda // To add new coming referenda to the end of list
+    }
+
+    // Remove items from cards if they were removed from basket
+    private fun checkItemsRemovedFromBasket(old: Map<ReferendumId, TinderGovBasketItem>, new: Map<ReferendumId, TinderGovBasketItem>) {
+        if (old.size > new.size) {
+            sortedReferendaFlow.value = emptyList()
+            _resetCards.sendEvent()
+        }
```

Comment:
~~I dont get his logic. Why are we removing everything if basket sice was reduced?~~

It is super unclear that we are removing everything here because `addNewReferendaToCards` will later add new data. This is requires a significant mental load to understand that

1. `checkItemsRemovedFromBasket` the name itself doesnt imply we basically reset everything
2. function "knows" about what happens on the upper level
3. It writes to the `sortedReferendaFlow` which is then read right afterwards in `addNewReferendaToCards` which works but is not ideal since it contradicts the pipeliening natue of Flows

Reply:
Okay, agree.
I think we can remove only items that doesn't exist in `new` list